### PR TITLE
airbyte-ci [experimental]: add pypi publishing

### PR DESCRIPTION
**Pull request by pedroslopez**
## What

We want to publish connectors to PyPI so they can be used for embedded connectors. This PR introduces an experimental command for publishing to pypi.

This is currently its own command and separate from the main publish pipelines until we validate that we really want to do this for all connectors to unblock the embedded connectors project.

## How

The approach taken here was very much to avoid modifying connectors directly. This leads us to have to do a couple things when publishing, to accomplish the following:
- Publish connectors following a common naming pattern, making sure they're prefixed by `airbyte-`, e.g. `airbyte-source-faker`
    - Our connectors tend to set the name as `source-faker` (sans `airbyte`) in the setup.py
- Publish to PyPI using the versioning information already present in the connector's metadata.yml
    - Our connectors tend to not set a version at all in the setup.py
- Ensure connectors are attributed to Airbyte as the author
    - Most are fine in this regard, but some have custom authors, e.g. `evan@airbyte.io` in the setup.py

To do the above without going through and changing all connectors, the following approach was taken:
1. Run the following steps within a dockerfile for environment isolation (makes it easier to run locally without changing a bunch of files in your filesystem)
2. Remove the name, version, and author fields from the setup.py if they exist. 
3. Add a setup.cfg with the desired name, version and author
4. Package and publish to PyPI

### CODE REVIEW

File: airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    async def _run(self) -> StepResult:
+        if self.context.connector.language not in ["python", "low-code"]:
+            return self.skip("Only Python connectors can be published to PyPI.")
+
+        try:
+            setup_cfg = dedent(
+                f"""
+                [metadata]
+                name = airbyte-{self.context.connector.technical_name}
```
Comment (flash1293) [inline airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py:26]: Does this mean the name of the package will be `airbyte-source-salesforce` ? I think that's fine, small note here that the module is still named `source_salesforce` (without the prefix). Probably not an issue, just calling it out

**Review by flash1293 — COMMENTED** [review]
This looks good, but not sure how to test it - should we just run a test release?

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py [hunk 2]
```diff
@@ -361,6 +362,75 @@ def publish(
...
+            dagger_logs_url=ctx.obj.get("dagger_logs_url"),
+            pipeline_start_timestamp=ctx.obj.get("pipeline_start_timestamp"),
+            ci_context=ctx.obj.get("ci_context"),
+            ci_gcs_credentials=ctx.obj["ci_gcs_credentials"],
+            pypi_username=pypi_username,
+            pypi_password=pypi_password,
+            pypi_repository=pypi_repository,
+        )
+        for connector, modified_files in selected_connectors_and_files.items()
+    ]
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:417]:
```suggestion
    connectors_contexts = [
        PublishPyPIContext(
            connector=connector,
            is_local=ctx.obj["is_local"],
            git_branch=ctx.obj["git_branch"],
            git_revision=ctx.obj["git_revision"],
            ci_report_bucket=ctx.obj["ci_report_bucket_name"],
            report_output_prefix=ctx.obj["report_output_prefix"],
            gha_workflow_run_url=ctx.obj.get("gha_workflow_run_url"),
            dagger_logs_url=ctx.obj.get("dagger_logs_url"),
            pipeline_start_timestamp=ctx.obj.get("pipeline_start_timestamp"),
            ci_context=ctx.obj.get("ci_context"),
            ci_gcs_credentials=ctx.obj["ci_gcs_credentials"],
            pypi_username=pypi_username,
            pypi_password=pypi_password,
            pypi_repository=pypi_repository,
        )
        for connector in ctx.obj["selected_connectors_with_modified_files"]
    ]
```

Following the changes of https://github.com/airbytehq/airbyte/pull/28855

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py [hunk 3]
```diff
@@ -361,6 +362,75 @@ def publish(
...
+
+    main_logger.warn("Concurrency is forced to 1. For stability reasons we disable parallel publish pipelines.")
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:419]:
```suggestion
    main_logger.warn("Concurrency is forced to 1. For stability reasons we disable parallel pypi publication pipelines.")
```

File: airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+from pipelines.bases import ConnectorReport, Step, StepResult, StepStatus
+from pipelines.contexts import PublishConnectorContext
+from pipelines.pipelines import metadata
+
+
+class PublishPyPIConnector(Step):
+    title = "Publish connector to PyPI"
+
+    async def _run(self) -> StepResult:
+        if self.context.connector.language not in ["python", "low-code"]:
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py:19]:
```suggestion
        if self.context.connector.language not in [ConnectorLanguage.PYTHON, ConnectorLanguage.LOW_CODE]:
```

File: airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            return self.skip("Only Python connectors can be published to PyPI.")
+
+        try:
+            setup_cfg = dedent(
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py:23]: Can we use jinja templates for templating `setup.py`?

File: airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                name = airbyte-{self.context.connector.technical_name}
+                version = {self.context.connector.version}
+                author = Airbyte
+                author_email = contact@airbyte.io
+            """
+            )
+
+            pypi_stdout = await (
+                self.context.dagger_client.container()
+                .from_("python:3.9-slim")
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py:35]: Is is mandatory to use python:3.9 here or could we publish python 3.9 packages from a python 3.10 env?

File: airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                .with_exec(["apt-get", "update"])
+                .with_exec(["apt-get", "install", "-y", "twine"])
+                .with_directory("connector", (await self.context.get_connector_dir(exclude=["dist"])))
+                .with_workdir("connector")
+                .with_exec(["sed", "-i", "/name=/d; /author=/d; /author_email=/d; /version=/d", "setup.py"])
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py:40]: What's the purpose of this `sed`? Can we provide a ready to consume `setup.py` to the container?

File: airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                .with_exec(["twine", "upload", "--verbose", "--repository", self.context.pypi_repository, "dist/*"])
+            ).stdout()
+
+            return StepResult(
+                self,
+                status=StepStatus.SUCCESS,
+                stdout=pypi_stdout,
+            )
+        except ExecError as e:
+            return StepResult(self, status=StepStatus.FAILURE, stderr=str(e))
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py:55]:
```suggestion
            twine_upload = await (
                self.context.dagger_client.container()
                .from_("python:3.9-slim")
                .with_exec(["apt-get", "update"])
                .with_exec(["apt-get", "install", "-y", "twine"])
                .with_directory("connector", (await self.context.get_connector_dir(exclude=["dist"])))
                .with_workdir("connector")
                .with_exec(["sed", "-i", "/name=/d; /author=/d; /author_email=/d; /version=/d", "setup.py"])
                .with_new_file("setup.cfg", setup_cfg)
                .with_exec(["pip", "install", "--upgrade", "setuptools", "wheel"])
                .with_exec(["python", "setup.py", "sdist", "bdist_wheel"])
                .with_env_variable("TWINE_USERNAME", self.context.pypi_username)
                .with_env_variable("TWINE_PASSWORD", self.context.pypi_password)
                .with_exec(["twine", "upload", "--verbose", "--repository", self.context.pypi_repository, "dist/*"])
            )

            return self.get_step_result(twine_upload)
        
```
Error handling and stdout parsing happens in `get_step_result`

File: airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                .with_exec(["apt-get", "update"])
+                .with_exec(["apt-get", "install", "-y", "twine"])
+                .with_directory("connector", (await self.context.get_connector_dir(exclude=["dist"])))
+                .with_workdir("connector")
+                .with_exec(["sed", "-i", "/name=/d; /author=/d; /author_email=/d; /version=/d", "setup.py"])
+                .with_new_file("setup.cfg", setup_cfg)
+                .with_exec(["pip", "install", "--upgrade", "setuptools", "wheel"])
+                .with_exec(["python", "setup.py", "sdist", "bdist_wheel"])
+                .with_env_variable("TWINE_USERNAME", self.context.pypi_username)
+                .with_env_variable("TWINE_PASSWORD", self.context.pypi_password)
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py:45]: Can you please use `with_secret_variable` to make sure Dagger obfuscate these from the logs if needed?
```python
twine_username_secret = self.context.dagger_client().set_secret("twine_username", self.context.pypi_username)
twine_password = self.context.dagger_client().set_secret("twine_password", self.context.pypi_password)

## Then when building the container you can
.with_secret_variable("TWINE_USERNAME", twine_username)
.with_secret_variable("TWINE_PASSWROD", twine_password)

```

**Review by alafanechere — CHANGES_REQUESTED** [review]
Could we make this part of the publish pipeline and add a `--publish-to-pypi` option on `publish` that defaults to false? It will make the upcoming refactor of the Context class easier 🙏

**Comment by alafanechere** [comment]
And please bump the package version 🙏 and update the doc with the new command / options.
Could you share in DM the context why we want to publish connectors to pypi?

File: airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            return self.skip("Only Python connectors can be published to PyPI.")
+
+        try:
+            setup_cfg = dedent(
+                f"""
+                [metadata]
+                name = airbyte-{self.context.connector.technical_name}
+                version = {self.context.connector.version}
+                author = Airbyte
+                author_email = contact@airbyte.io
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py:29]: If it's a community contributed connector could we keep the author and author email?

File: airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py (hunk 6 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py:35]: Would be team 3.10 if possible!

**Comment by aaronsteers** [comment]
Notes from conversation today with @flash1293 and @pedroslopez.

Regarding `setup.cfg`:
1. Package name conformance: airbyte-<connector-type>-<connector-name> instead of `source-hubspot`.
2. Version is not set. This sets it.
3. Standardize 'author'.

Todo on me:

- Check if setup.cfg is compatible with Poetry?

Re: setup.cfg contents: https://github.com/airbytehq/airbyte/blob/5a3962a6c9a764e4ade8484a7e6f0ca88eda1c82/airbyte-ci/connectors/pipelines/pipelines/publish_pypi.py#L22-L29

**Comment by aaronsteers** [comment]
Another TODO:

- Is there a way to standardize how version is determined and retrieved?

**Pull request closed without merging**